clients/src/main/java/org/apache/kafka/common/security/kerberos/Login.java

```diff
@@ -0,0 +1,373 @@
+                                + " Exiting refresh thread.", nextRefreshDate);
+                        return;
+                    }
+                    if (isUsingTicketCache) {
+                        String kinitArgs = "-R";
+                        int retry = 1;
+                        while (retry >= 0) {
+                            try {
+                                log.debug("Running ticket cache refresh command: {} {}", kinitCmd, kinitArgs);
+                                Shell.execCommand(kinitCmd, kinitArgs);
+                                break;
+                            } catch (Exception e) {
+                                if (retry > 0) {
+                                    --retry;
+                                    // sleep for 10 seconds
+                                    try {
+                                        Thread.sleep(10 * 1000);
+                                    } catch (InterruptedException ie) {
+                                        log.error("Interrupted while renewing TGT, exiting Login thread");
+                                        return;
+                                    }
+                                } else {
+                                    log.warn("Could not renew TGT due to problem running shell command: '" + kinitCmd
+                                            + " " + kinitArgs + "'" + "; exception was: " + e + ". Exiting refresh thread.", e);
+                                    return;
+                                }
+                            }
+                        }
+                    }
+                    try {
+                        int retry = 1;
+                        while (retry >= 0) {
+                            try {
+                                reLogin();
+                                break;
+                            } catch (LoginException le) {
+                                if (retry > 0) {
+                                    --retry;
+                                    // sleep for 10 seconds.
+                                    try {
+                                        Thread.sleep(10 * 1000);
+                                    } catch (InterruptedException e) {
+                                        log.error("Interrupted during login retry after LoginException:", le);
+                                        throw le;
+                                    }
+                                } else {
+                                    log.error("Could not refresh TGT for principal: " + principal + ".", le);
+                                }
+                            }
+                        }
+                    } catch (LoginException le) {
+                        log.error("Failed to refresh TGT: refresh thread exiting now.", le);
+                        return;
+                    }
+                }
+            }
+        }, false);
```

Comment:
@junrao suggested that it shouldn't be because we wait for it to terminate on `shutdown`. And if consumers are closed, then it won't prevent client applications from terminating. But it may cause this problem when consumers are not closed, so I am tempted to change it back to a daemon thread. What are your thoughts Jun?